src/boot/efi/console.c

```diff
@@ -25,49 +43,74 @@ EFI_STATUS console_key_read(UINT64 *key, BOOLEAN wait) {
                 if (EFI_ERROR(err))
                         TextInputEx = NULL;
 
+                err = uefi_call_wrapper(BS->CheckEvent, 1, TextInputEx->WaitForKeyEx);
+                if (err == EFI_INVALID_PARAMETER)
+                        /* The firmware pretends it talks this protocol, but it really doesn't. */
+                        TextInputEx = NULL;
+                else
+                        events[event_count++] = TextInputEx->WaitForKeyEx;
+
                 checked = TRUE;
         }
 
-        /* wait until key is pressed */
-        if (wait)
-                uefi_call_wrapper(BS->WaitForEvent, 3, 1, &ST->ConIn->WaitForKey, &index);
+        if (timeout_usec > 0) {
+                err = uefi_call_wrapper(BS->CreateEvent, 5, EVT_TIMER, 0, NULL, NULL, &timer);
+                if (EFI_ERROR(err))
+                        return log_error_status_stall(err, L"Error creating timer event: %r");
 
-        if (TextInputEx) {
+                /* SetTimer expects 100ns units for some reason. */
+                err = uefi_call_wrapper(BS->SetTimer, 3, timer, TimerRelative, timeout_usec * 10);
+                if (EFI_ERROR(err)) {
+                        uefi_call_wrapper(BS->CloseEvent, 1, timer);
```

Comment:
maybe we should add a `_cleanup_` helper for CloseEvent, instead of doing the clean-up manually in all error paths.

Reply:
So far it's only those two occasions. Seems more hassle than it's worth for some efi opaque type.